plugins/modules/foreman_smart_class_parameter.py

```diff
@@ -0,0 +1,248 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# (c) 2020 Baptiste Agasse (@bagasse)
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: foreman_smart_class_parameter
+short_description: Manage Foreman Smart Class Parameters using Foreman API
+description:
+  - Update Foreman Smart Class Parameters using Foreman API. Smart Class Paramters are created/deleted at on puppet classes import
+  - and cannot be created or deleted via Foreman API.
+author:
+  - "Baptiste Agasse (@bagasse)"
+options:
+  puppetclass_name:
+    description: Name of the puppetclass that own the parameter
+    required: true
+    type: str
+  parameter:
+    description: Name of the parameter
+    required: true
+    type: str
+  description:
+    description: Description of the Smart Class Parameter
+    type: str
+  override:
+    description: Whether the smart class parameter value is managed by Foreman
+    type: bool
+  default_value:
+    description: Value to use by default.
+    type: raw
+  hidden_value:
+    description: When enabled the parameter is hidden in the UI.
+    type: bool
+  omit:
+    description:
+      - Foreman will not send this parameter in classification output.
+      - Puppet will use the value defined in the Puppet manifest for this parameter.
+    type: bool
+  override_value_order:
+    description: The order in which values are resolved.
+    type: list
+    elements: str
+  validator_type:
+    description: Types of validation values.
+    type: str
+    choices:
+      - regexp
+      - list
+  validator_rule:
+    description: Used to enforce certain values for the parameter values.
+    type: str
+  parameter_type:
+    description: Types of variable values. Set an empty to reset the parameter type.
+    type: str
+    choices:
+      - string
+      - boolean
+      - integer
+      - real
+      - array
+      - hash
+      - yaml
+      - json
+      - ''
```

Comment:
I wonder if this rather should be "reset" or something. I know the Foreman API wants an empty string (or rather `nil`), but this feels weird from a UX perspective.

Reply:
Yes, it will be more readable if we use `reset` or `none` instead of empty string. Any preference between `reset` or `none` ?

Reply:
I think this depends on the behavior of the module/api. I think `none` is fine if the action is idempotent (aka the reply of a show action later will also yield "" or `nil` or an unset parameter). And `reset` should be used if the next show action will show us `string` or whatever. Does it make sense?

Also, how does our param to string helper handle this?

Reply:
Ok, moved this to `none`, it fallback to string if empty for parm to string helper